mlflow/spark/__init__.py

```diff
@@ -992,27 +989,50 @@ def predict(
+        # otherwise it will break `spark.createDataFrame` column type inferring.
+        if self.signature and self.signature.inputs:
+            for col_spec in self.signature.inputs.inputs:
+                if isinstance(col_spec.type, SparkMLVector):
+                    col_name = col_spec.name or pandas_df.columns[0]
+
+                    pandas_df[col_name] = pandas_df[col_name].map(
+                        lambda array: [float(elem) for elem in array]
+                    )
+
+        spark_df = self.spark.createDataFrame(pandas_df)
+
+        # Convert Array[Double] column to spark ML vector type according to signature
+        if self.signature and self.signature.inputs:
+            for col_spec in self.signature.inputs.inputs:
+                if isinstance(col_spec.type, SparkMLVector):
+                    from pyspark.ml.functions import array_to_vector
+
+                    col_name = col_spec.name or spark_df.columns[0]
+                    spark_df = spark_df.withColumn(col_name, array_to_vector(col_name))
+
+        # For the case of no signature or signature logged by old version MLflow,
+        # the signature does not support spark ML vector type, in this case,
+        # automatically infer vector type input columns and do the conversion
+        # using `_find_and_set_features_col_as_vector_if_needed` utility function.
+        spark_df = _find_and_set_features_col_as_vector_if_needed(spark_df, self.spark_model)
+
+        prediction_column = mlflow.pyspark.ml._check_or_set_model_prediction_column(
+            self.spark_model, spark_df
         )
-        prediction_column = "prediction"
-        if isinstance(self.spark_model, PipelineModel) and self.spark_model.stages[-1].hasParam(
-            "outputCol"
-        ):
-            from pyspark.sql import SparkSession
-
-            spark = SparkSession.builder.getOrCreate()
-            # do a transform with an empty input DataFrame
-            # to get the schema of the transformed DataFrame
-            transformed_df = self.spark_model.transform(spark.createDataFrame([], spark_df.schema))
-            # Ensure prediction column doesn't already exist
-            if prediction_column not in transformed_df.columns:
-                # make sure predict work by default for Transformers
-                self.spark_model.stages[-1].setOutputCol(prediction_column)
-        return [
-            x.prediction
-            for x in self.spark_model.transform(spark_df).select(prediction_column).collect()
-        ]
+        prediction_df = self.spark_model.transform(spark_df).select(prediction_column)
+
+        # If signature output schema exists and it contains vector type columns,
+        # Convert spark ML vector type column to Array[Double] otherwise it will
+        # break enforce_schema checking
+        if self.signature and self.signature.outputs:
+            for col_spec in self.signature.outputs.inputs:
```

Comment:
(not comment but just a random complaint for sth implemented years ago) the `outputs.inputs` looks very unnatural😂

Reply:
yes this is strange, but it is another problem. we can improve it in follow-up PR.